### PR TITLE
Optimisation for memory when constructing param_grid in grid_search

### PR DESCRIPTION
**Pull request by digital10111**
Earlier param_grid was being constructed with the train and test indices against a parameter. This would increase the memory consumption and cause OMem in cases where the number of instances were large. 
Changed the param_grid construction so that only test sequence number resides with the parameters. 

The actual train and test dataset is easily retrieved in the executor. Thereby reducing the memory consumption in the driver.

### CODE REVIEW

**Comment by digital10111** [comment]
@jkbradley Hi, the tests seems to fail with` Intel MKL FATAL ERROR: Cannot load libmkl_avx2.so or libmkl_def.so.`

Doesn't really seems to be an issue with the changes I have made.

Help would be appreciated.

File: python/spark_sklearn/grid_search.py [hunk 1]
```diff
@@ -308,10 +308,11 @@ def _fit(self, X, y, groups, parameter_iterable):
         fas = _fit_and_score
 
         def fun(tup):
-            (index, (parameters, train, test)) = tup
+            (index, (parameters, test_sequence_index)) = tup
             local_estimator = clone(base_estimator)
             local_X = X_bc.value
             local_y = y_bc.value
+            train, test = next(islice(cv.split(local_X, local_y, groups), test_sequence_index, test_sequence_index + 1))
```
Comment (srowen) [inline python/spark_sklearn/grid_search.py]: Shall we broadcast `groups` as well here? it is also potentially as long as the size of the data set.

Pardon the basic questions as I'm not as familiar with Python. I was wondering why you'd get an islice just to get one element, but is the point that cv.split returns a generator and so this is, at least, memory-efficient?

It would cause every task to re-split the data, which could be slower, right? I also wonder if this will produce the same splits every time, as it can shuffle the data.

I see the problem, that a whole set of indices is stored per element of the param grid map. There's no real need to do that anyway as we just need to run all pairs of (train, test) splits with all param_grid elements. 

I think you could match all pairs with Spark with a .cartesian of two RDDs. The trick would be to try to reconstruct the exact same index, but that's just a matter of supplying an index with each RDD and some math.

Or simply have `fun` run all (train, test) splits for an individual grid element. That batches work together more, which is simple and probably faster, but bad if you have a lot of parallelism but few param grid elements, and many folds.

WDYT? if you are interested in trying it, go for it.

Comment (digital10111) [reply python/spark_sklearn/grid_search.py]: > Shall we broadcast `groups` as well here? it is also potentially as long as the size of the data set.

Yes, I think we should. 

> the point that cv.split returns a generator and so this is, at least, memory-efficient?

You are correct.

> It would cause every task to re-split the data, which could be slower, right?

Yes, it would cause every task to re-split the data.  In my experiments re-splitting the data in each task, wasn't that computationally heavy and didn't slow down the tasks. But since you have raised a valid concern, I shall profile the split method.

> I also wonder if this will produce the same splits every time, as it can shuffle the data.

For splitters which are **not ShuffleSplitters**, it will always have the same set of indices, across tasks. 
For splitters **which are ShuffleSplitters**, [we can set an integer `random_state`](https://github.com/scikit-learn/scikit-learn/blob/55bf5d93e5674f13a1134d93a11fd0cd11aabcd1/sklearn/model_selection/_split.py#L407), to maintain the splits across tasks. [[`GroupShuffleSplit`, `ShuffleSplit`, `StratifiedShuffleSplit`]](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.model_selection)

> Or simply have `fun` run all (train, test) splits for an individual grid element. That batches work together more, which is simple and probably faster, but bad if you have a lot of parallelism but few param grid elements, and many folds.

I have tried it, it was slower and a lot more unnatural in terms of Apis as well. Sending split level fit calls to executors was much faster and more natural in terms of writing code.



> I think you could match all pairs with Spark with a .cartesian of two RDDs. The trick would be to try to reconstruct the exact same index, but that's just a matter of supplying an index with each RDD and some math.

Sounds quite hand wavy. As far matching all pairs and reconstructing the exact same index. I believe that is being achieved anyway, except for ShuffleSplitters, which can be easily solved setting the `random_state` to an integer.

Comment (digital10111) [reply python/spark_sklearn/grid_search.py]: one hundred thousand
```
shape_x = 10**5
shape_y = 10*2
X = np.random.rand(shape_x,shape_y)
y = np.random.randint(0, 2, (shape_x, ))
skf = ShuffleSplit(n_splits=10, random_state=42)
%timeit next(islice(skf.split(X, y), 9, 10))
100 loops, best of 3: 19.4 ms per loop
```
one million
```
shape_x = 10**6
shape_y = 10*2
X = np.random.rand(shape_x,shape_y)
y = np.random.randint(0, 2, (shape_x, ))
skf = ShuffleSplit(n_splits=10, random_state=42)
%timeit next(islice(skf.split(X, y), 9, 10))
1 loop, best of 3: 227 ms per loop
```
ten million
```
shape_x = 10**7
shape_y = 10*2
X = np.random.rand(shape_x,shape_y)
y = np.random.randint(0, 2, (shape_x, ))
skf = ShuffleSplit(n_splits=10, random_state=42)
%timeit next(islice(skf.split(X, y), 9, 10))
1 loop, best of 3: 4.7 s per loop
```
one hundred million

```
shape_x = 10**8
shape_y = 10*2
X = np.random.rand(shape_x,shape_y)
y = np.random.randint(0, 2, (shape_x, ))
skf = ShuffleSplit(n_splits=10, random_state=42)
%timeit next(islice(skf.split(X, y), 9, 10))
1 loop, best of 3: 1min 4s per loop
```

Comment (srowen) [reply python/spark_sklearn/grid_search.py]: The task-per-param-and-fold Spark option isn't hard, but in the end I suspect it's slower than your current approach because Spark has to shuffle a bunch of data. Off the top of my head, it's just this (in Scala, easier for me to write off the top of my head) for what it's worth:

```
val numParams = parameter_iterable.size()
val indexedTrainTest = sc.parallelize(cv.split(X, y, groups).zipWithIndex()
val indexedParams = sc.parallelize(parameter_iterable).zipWithIndex()
indexedTrainTest.cartesian(indexedParams).map { case (((train, test), t), (param, p)) =>
  val index = (t * numParams + p, (param, train, test))
  ...
}
```

The task-per-param Spark version where all folds are considered in one task also isn't hard:

```
val numParams = parameter_iterable.size()
val indexedParams = sc.parallelize(parameter_iterable).zipWithIndex()
val trainTestBroadcast = sc.broadcast(cv.split(X, y, groups).zipWithIndex())
indexedParams.flatMap { case (param, p) =>
  trainTestBroadcast.value.map { case ((train, test), t) =>
      val index = (t * numParams + p, (param, train, test))
      ...
  }
}
```

I'm OK sticking to your approach right now as it doesn't change the distribution of the data and maximizes parallelism, at the cost of some overhead.

If re-splitting the data adds, say, 5 seconds when the data set is about 10m records, then that's non-trivial but still small relative to the overhead of Spark and the model fitting itself, which probably runs at least tens of seconds at that scale.

Yeah we do need to fix the `random_state`. `GridSearchCV` here and in scikit don't seem to let you fix a seed. Therefore I think it's fine to pick a random seed in this method, but then apply it to all calls to cv.split.

Comment (digital10111) [reply python/spark_sklearn/grid_search.py]: > val indexedTrainTest = sc.parallelize(cv.split(X, y, groups).zipWithIndex() 
> val indexedParams = sc.parallelize(parameter_iterable).zipWithIndex() 
> indexedTrainTest.cartesian(indexedParams).map

Although we are keeping the splits in the driver's memory, this not as bad as the current master. And does exactly what this PR does, except for the overhead of shuffling the data as you mentioned. I might still be slightly wary about the memory pressure on the driver.

> Yeah we do need to fix the `random_state`. `GridSearchCV` here and in scikit don't seem to let you fix a seed.

That's right, I have made a fix for the same. Would set `random_state` in the cv object to an integer in the driver itself, if it's not already. Which would be propagated to all tasks as is.

Comment (digital10111) [reply python/spark_sklearn/grid_search.py]: > > val indexedTrainTest = sc.parallelize(cv.split(X, y, groups).zipWithIndex()
> > val indexedParams = sc.parallelize(parameter_iterable).zipWithIndex()
> > indexedTrainTest.cartesian(indexedParams).map
> 
> Although we are keeping the splits in the driver's memory, this not as bad as the current master, and does exactly what this PR does. Except for the overhead of shuffling the data as you mentioned. I might still be slightly wary about the memory pressure on the driver.
> 
> > Yeah we do need to fix the `random_state`. `GridSearchCV` here and in scikit don't seem to let you fix a seed.
> 
> That's right, I have made a fix for the same. Would set `random_state` in the cv object to an integer in the driver itself, if it's not already. Which would be propagated to all tasks as is. 
Would it matter too much, if we didn't set the random_state,  statistically? Just a small difference in estimates of the out of sample error, across different params.

**Comment by srowen** [comment]
I'm OK with it. I wonder if @thunterdb has thoughts on it.

File: python/spark_sklearn/grid_search.py [hunk 2]
```diff
@@ -277,7 +278,11 @@ def _fit(self, X, y, groups, parameter_iterable):
 
         estimator = self.estimator
         cv = check_cv(self.cv, y, classifier=is_classifier(estimator))
-
+        
+        if hasattr(cv, 'random_state'):
```
Comment (thunterdb) [inline python/spark_sklearn/grid_search.py]: why is this necessary?

Comment (srowen) [reply python/spark_sklearn/grid_search.py]: Not all of the CV split implementations have a random element to them

**Comment by thunterdb** [comment]
LGTM

File: python/spark_sklearn/grid_search.py (hunk 2 above)
Comment (digital10111) [reply python/spark_sklearn/grid_search.py]: Since, "An iterable yielding train/test splits" is a legal argument to `check_cv`. I believe one might be able to write an implementation of cv without a `random_state`.  An example could be [LeaveOneOut](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.LeaveOneOut.html).
